MekHQ/src/mekhq/campaign/stratcon/StratconContractInitializer.java

```diff
@@ -0,0 +1,340 @@
+     */
+    private static void initializeTrackFacilities(StratconTrackState trackState, int numFacilities, ForceAlignment owner, 
+            List<String> facilitySubset, boolean strategicObjective) {
+        for (int fCount = 0; fCount < numFacilities; fCount++) {
+            StratconFacility sf =
+                    facilitySubset == null ? 
+                    StratconFacilityFactory.getRandomFacility() :
+                    StratconFacilityFactory.getFacilityByName(facilitySubset.get(Compute.randomInt(facilitySubset.size())));
+            sf.setOwner(owner);
+            sf.setStrategicObjective(strategicObjective);
+            
+            int x = Compute.randomInt(trackState.getWidth());
+            int y = Compute.randomInt(trackState.getHeight());
+            StratconCoords coords = new StratconCoords(x, y);
+            
+            // make sure we don't put the facility down on top of anything else
+            while ((trackState.getFacility(coords) != null) ||
+                    (trackState.getScenario(coords) != null)) {
+                x = Compute.randomInt(trackState.getWidth());
+                y = Compute.randomInt(trackState.getHeight());
+                coords = new StratconCoords(x, y);
+            }
+            
+            trackState.addFacility(coords, sf);
+            if (sf.getOwner() == ForceAlignment.Allied) {
+                trackState.getRevealedCoords().add(coords);
+                sf.setVisible(true);
+            } else {
+                sf.setVisible(false);
+            }
+        }
+    }
+    
+    /**
+     * Worker function that takes a trackstate and plops down the given number of facilities owned by the given faction
+     */
+    private static void initializeObjectiveScenarios(Campaign campaign, AtBContract contract, StratconTrackState trackState, 
+            int numScenarios, List<String> objectiveScenarios, List<String> objectiveModifiers) {
+        // pick scenario from subset
+        // place it on the map somewhere nothing else has been placed yet
+        // if it's a facility scenario, place the facility
+        // run generateScenario() to apply all the necessary mods
+        // apply objective mods (?)
+        
+        
+        for (int sCount = 0; sCount < numScenarios; sCount++) {
+            // pick
+            ScenarioTemplate template = StratconScenarioFactory.getSpecificScenario(
+                    objectiveScenarios.get(Compute.randomInt(objectiveScenarios.size())));
+            
+            // plonk
+            int x = Compute.randomInt(trackState.getWidth());
+            int y = Compute.randomInt(trackState.getHeight());
+            StratconCoords coords = new StratconCoords(x, y);
+            
+            // make sure we don't put the facility down on top of anything else
+            while ((trackState.getFacility(coords) != null) ||
+                    (trackState.getScenario(coords) != null)) {
```

Comment:
Infinite loop? Also, given this is used multiple times it might be best to create a separate method for this.